muspy/base.py

```diff
@@ -652,43 +662,28 @@ def remove_invalid(
         Object itself.
 
         """
-        if attr is None:
-            for attribute in self._list_attributes:
-                self._remove_invalid(attribute, recursive)
-        elif attr in self._list_attributes:
-            self._remove_invalid(attr, recursive)
-        else:
-            raise TypeError("`{}` must be a list attribute.")
-        return self
-
-    def _remove_duplicate(self, attr: str, recursive: bool):
-        # Skip it if empty
-        if not getattr(self, attr):
-            return
+        # NOTE: We depend on the fact that _traverse_lists uses
+        # post-order traversal. This way, we first apply recursively
+        # and later check the currect object so that something that
+        # can be fixed in a lower level would not make the high-level
+        # object to be removed.
+
+        attr_type: Type
+        value: list
+        for _, attr_type, value in self._traverse_lists(attr, recursive):
+            # Skip it if empty
+            if not value:
+                continue
 
-        attr_type = self._attributes[attr]
-        is_complexbase = isclass(attr_type) and issubclass(
-            attr_type, ComplexBase
-        )
-        value = getattr(self, attr)
-
-        # NOTE: The ordering mathers here. We first apply recursively
-        # and later check the currect object so that something that can
-        # be fixed in a lower level would not make the high-level object
-        # removed.
-
-        # Apply recursively
-        if recursive and is_complexbase:
-            for item in value:
-                item.remove_duplicate(recursive=recursive)
+            if isclass(attr_type) and issubclass(attr_type, Base):
+                value[:] = [item for item in value  # type: ignore
+                            if (isinstance(item, attr_type)
+                                and cast(Base, item).is_valid())]
+            else:
+                value[:] = [item for item in value  # type: ignore
+                            if isinstance(item, attr_type)]
```

Comment:
Type checking can be simplified if we change the return signature into `Iterator[Tuple[str, List]]`.

Reply:
I think there must be a bug in mypy, it was complaining about things whose type is actually known...